### PR TITLE
macro kwarg escaping fix wip

### PR DESCRIPTION
**Pull request by wsphillips**
Not exactly fixed...but close I think @jpsamaroo have a look? Likely needs just a minor tweak because of the way we are handling AMD-specific arguments. Mirroring the fix Tim made [here](https://github.com/JuliaGPU/CUDAnative.jl/commit/0bc69cff6f2ab5a1a1d13d08e30d5cfe22dddab0#diff-8c885cfac192f9fa48c582d66d0ea06c)

### CODE REVIEW

**Comment by jpsamaroo** [comment]
@maleadt is the fix that Wiktor pointed out above just a workaround for a bug in Julia 1.5, or was there a change in Julia's behavior that we're expected to account for? IMO this change seems to make the code look a lot uglier and less intuitive.

**Comment by maleadt** [comment]
See https://github.com/JuliaLang/julia/issues/23221#issuecomment-321953292... IIUC it's been broken for a long time, but I didn't run into that with how the `@cuda` macro was structured up until recently with Julia 1.5.

**Comment by jpsamaroo** [comment]
Unfortunate, but there's not much else we can do here... Thanks for the heads-up! :smile:

File: src/execution.jl [hunk 1]
```diff
@@ -138,21 +138,27 @@ macro roc(ex...)
...
-                local device = extract_device(; $(map(esc, call_kwargs)...))
-                local kernel = rocfunction(device, $(esc(f)), kernel_tt;
-                                           $(map(esc, compiler_kwargs)...))
-                local queue = extract_queue(device; $(map(esc, call_kwargs)...))
-                local signal = create_event()
-                kernel(queue, signal, kernel_args...; $(map(esc, call_kwargs)...))
-                wait(signal)
+                local $kernel_args = map(rocconvert, ($(var_exprs...),))
+                local $kernel_tt = Tuple{Core.Typeof.($kernel_args)...}
+                local $device = AMDGPUnative.extract_device(; $(call_kwargs...))
```
Comment (maleadt) [inline src/execution.jl]: You can also do `$extract_device` etc to resolve the function eagerly.

Comment (jpsamaroo) [reply src/execution.jl]: Will push this with my next commit.

**Comment by jpsamaroo** [comment]
So I spent a little while trying to tear this apart, and so far the only change I can see between Julia 1.3 and Julia master is the following difference in the binary output by LLVM (shown with `nm`):

Julia 1.3:
```
0000000000000000 R __oclc_ISA_version
000000000000000c R __oclc_correctly_rounded_sqrt32
0000000000000010 R __oclc_daz_opt
0000000000000004 R __oclc_finite_only_opt
0000000000000008 R __oclc_unsafe_math_opt
0000000000000000 i roccall_vadd_28
```

Julia master:
```
0000000000000000 R __oclc_ISA_version
000000000000000c R __oclc_correctly_rounded_sqrt32
0000000000000010 R __oclc_daz_opt
0000000000000004 R __oclc_finite_only_opt
0000000000000008 R __oclc_unsafe_math_opt
0000000000000000 T roccall_vadd_28
0000000000000000 R roccall_vadd_28.kd
```

**Comment by jpsamaroo** [comment]
Output of `readelf -s`:

Julia 1.3:
```
Symbol table '.symtab' contains 7 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
     1: 0000000000000000     4 OBJECT  GLOBAL DEFAULT    4 __oclc_ISA_version
     2: 000000000000000c     4 OBJECT  GLOBAL DEFAULT    4 __oclc_correctly_rounded_
     3: 0000000000000010     4 OBJECT  GLOBAL DEFAULT    4 __oclc_daz_opt
     4: 0000000000000004     4 OBJECT  GLOBAL DEFAULT    4 __oclc_finite_only_opt
     5: 0000000000000008     4 OBJECT  GLOBAL DEFAULT    4 __oclc_unsafe_math_opt
     6: 0000000000000000   368 <OS specific>: 10 GLOBAL DEFAULT    2 roccall_vadd_28
```

Julia master:
```
Symbol table '.symtab' contains 9 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
     1: 0000000000000000     0 SECTION LOCAL  DEFAULT    2
     2: 0000000000000000     4 OBJECT  GLOBAL DEFAULT    5 __oclc_ISA_version
     3: 000000000000000c     4 OBJECT  GLOBAL DEFAULT    5 __oclc_correctly_rounded_
     4: 0000000000000010     4 OBJECT  GLOBAL DEFAULT    5 __oclc_daz_opt
     5: 0000000000000004     4 OBJECT  GLOBAL DEFAULT    5 __oclc_finite_only_opt
     6: 0000000000000008     4 OBJECT  GLOBAL DEFAULT    5 __oclc_unsafe_math_opt
     7: 0000000000000000   112 FUNC    GLOBAL PROTECTED    2 roccall_vadd_28
     8: 0000000000000000    64 OBJECT  GLOBAL DEFAULT    3 roccall_vadd_28.kd
```

**Comment by jpsamaroo** [comment]
This PR is doing what it's intended to do, so I'm going to merge this and continue investigating the HSA symbol error separately. Thanks @wsphillips !

**Pull request merged**